src/compiler/declarationEmitter.ts

```diff
@@ -133,6 +176,419 @@ namespace ts {
             referencePathsOutput,
         };
 
+        function emitReferences(sourceFile: SourceFile, emittedReferencedFiles: FileReference[]) {
+            // Check what references need to be added
+            if (!compilerOptions.noResolve) {
+                forEach(sourceFile.referencedFiles, fileReference => {
+                    let referencedFile = tryResolveScriptReference(host, sourceFile, fileReference);
+
+                    // If the reference file is a declaration file or an external module, emit that reference
+                    if (referencedFile && (isExternalModuleOrDeclarationFile(referencedFile) &&
+                        !contains(emittedReferencedFiles, referencedFile))) { // If the file reference was not already emitted
+
+                        writeReferencePath(referencedFile);
+                        emittedReferencedFiles.push(referencedFile);
+                    }
+                });
+            }
+        }
+
+        function emitFlattenedTypeDefinitions(entrypoint: SourceFile): void {
+            let undoActions: (() => void)[] = [];
+            let aliasEmits: (() => void)[] = [];
+            let symbolNameSet: Map<number> = {};
+            let imports: Map<Declaration[]> = {};
+            let importEquals: Map<Declaration[]> = {};
+            let exportedMembers = resolver.getExportsOfModule(entrypoint.symbol);
+
+            // Handle an export=import as soon as possible
+            let maybeRedirectionType = exportedMembers.length && resolver.getDefiningTypeOfSymbol(exportedMembers[0]);
+            if (maybeRedirectionType && maybeRedirectionType.symbol && exportedMembers[0].name === "export=" && maybeRedirectionType.symbol.valueDeclaration && maybeRedirectionType.symbol.valueDeclaration.kind === SyntaxKind.SourceFile) {
+                return emitFlattenedTypeDefinitions(maybeRedirectionType.symbol.valueDeclaration as SourceFile);
+            }
+
+            let createSyntheticIdentifiers = (symbol: Symbol, generatedName: string) => {
+                forEach(symbol.declarations, declaration => {
+                    let synthId = createSynthesizedNode(SyntaxKind.Identifier) as Identifier;

```

Comment:
look aside table instead. the first thing is to have an emitIdentifier routine that all methods writing names would use, then this would be the only place to check the lookaside table for the replacement string.